### PR TITLE
[IDE/SourceKit] New SourceKit request for filtered method list

### PR DESCRIPTION
**Pull request by rintaro**
`source.request.conformingmethods` is a new SourceKit request which receives a source position and a list of protocol namses, returns a list of methods whose return type conforms to the requested protocols.

rdar://problem/44699573

### CODE REVIEW

**Comment by rintaro** [comment]
@swift-ci Please smoke test

File: lib/IDE/ConformingMethodList.cpp [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+}
+
+void ConformingMethodListCallbacks::resolveExpectedTypes(
+    ArrayRef<const char *> names, SourceLoc loc,
+    SmallVectorImpl<ProtocolDecl *> &result) {
+  auto &ctx = CurDeclContext->getASTContext();
+
+  for (auto name : names) {
+    std::string err;
+    if (auto D = getDeclFromMangledSymbolName(ctx, name, err)) {
```
[the remark this answers is not in the document]
Comment (slavapestov) [reply lib/IDE/ConformingMethodList.cpp]: Yes, let’s not use this API. Instead, you should be able to call Demangle::getTypeForMangling() (you might have to adjust the inputs that are passed in to be mangled types and not mangled decls, I think they differ slightly for protocols)

File: lib/IDE/ConformingMethodList.cpp [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+      : CodeCompletionCallbacks(P), ExpectedTypeNames(ExpectedTypeNames),
+        Consumer(Consumer) {}
+
+  // Only handle callbacks for suffix completions.
+  // {
+  void completeDotExpr(Expr *E, SourceLoc DotLoc) override;
+  void completePostfixExpr(Expr *E, bool hasSpace) override;
+  // }
+
+  // Ignore other callbacks.
```
[the remark this answers is not in the document]
Comment (rintaro) [reply lib/IDE/ConformingMethodList.cpp]: That's good idea! Will do.

File: lib/IDE/ConformingMethodList.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (rintaro) [reply lib/IDE/ConformingMethodList.cpp]: OK, I'll do that. The current expected input is mangled *types* anyway.

**Comment by rintaro** [comment]
@swift-ci Please smoke test

File: lib/IDE/ConformingMethodList.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (slavapestov) [reply lib/IDE/ConformingMethodList.cpp]: Thank you!

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Pull request merged**